imports/plugins/core/orders/server/no-meteor/mutations/placeOrder.js

```diff
@@ -261,6 +262,27 @@ export default async function placeOrder(context, input) {
     }
   };
 
+  let referenceId;
+  // Grabbing cart to get the cart.referenceId. It's a shame since we are fetch cart twice before
+  const cart = await Cart.findOne({ _id: cartId }, { _id: 0, referenceId: 1});
```

Comment:
I would move this `findOne` to above the discounts block, and update `getDiscountsTotalForCart` function to have the `cart` passed in rather than just the cart ID. `getDiscountsTotalForCart` is called only in two places, and in the other place (discounts startup code), the full `cart` is already available. So this change will prevent a duplicate cart lookup in two places.

Also, you could check but I don't think `_id: 0` does anything when combined with other fields that are `1`. They may have changed this in more recent Mongo versions.

Reply:
Two more things:

Add this after the `findOne` line:

```js
if (!cart) {
  throw new ReactionError("not-found", "Cart not found");
}
```

And `cartId` isn't required so you need to check it. If falsy, you can skip the `findOne`. And do something like `if (cart && cart.referenceId)` below, or somehow make sure `cart` is defined.

Reply:
@aldeed I had considered just leveraging the `getDiscountsTotalsForCart` but it felt like a bad separations of concerns, and somebody could eliminate or change that method and not realize the side effects. It seems like the better change would be to grab the cart and modify that method to take a passed in cart.

Reply:
I don't understand, how is `cartId` not required here

Reply:
As far as I know `_id: 0` has worked with every version of Mongo I have used at least
![2019-03-20_10-10-31](https://user-images.githubusercontent.com/72819/54654231-72360b00-4af8-11e9-9846-904ca3675807.png)

Reply:
You're right. I just tried it. I know there was an early version where if you had any `1` fields, the `0` fields were ignored. But I don't think I've tested it since 5 or 6 years ago.